share/eds/inverted_index.go

```diff
@@ -0,0 +1,92 @@
+package eds
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/filecoin-project/dagstore/index"
+	"github.com/filecoin-project/dagstore/shard"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"github.com/multiformats/go-multihash"
+)
+
+// simpleInvertedIndex is an inverted index that only stores a single shard key per multihash. Its
+// implementation is modified from the default upstream implementation in dagstore/index.
+type simpleInvertedIndex struct {
+	mu sync.Mutex
+	ds ds.Batching
+}
+
+// newSimpleInvertedIndex returns a new inverted index that only stores a single shard key per
+// multihash. This is because we use badger as a storage backend, so updates are expensive, and we
+// don't care which shard is used to serve a cid.
+func newSimpleInvertedIndex(dts ds.Batching) *simpleInvertedIndex {
+	dts = namespace.Wrap(dts, ds.NewKey("/inverted/index"))
+	return &simpleInvertedIndex{
+		ds: dts,
+	}
+}
+
+func (s *simpleInvertedIndex) AddMultihashesForShard(
+	ctx context.Context,
+	mhIter index.MultihashIterator,
+	sk shard.Key,
+) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
Ok. the original impl also has it. I guess because multiple Iterators may overlap in keys, i.e, when two iterators are writing simultaneously, the last one will overwrite the previous without catching indices the previous brought 
In our case, this is not the case. The only race which may happen is who writes first, IIUC, so we may not need this lock

Reply:
We need this lock to avoid updating the key. Technically it is unnecessary because the amount of times this occurs won't actually bloat the value log, so removing it will speed up indexing at the cost of a little bit of wie amplification